Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -889,6 +890,12 @@ def _check_valid_cython_module(self, pos, module_name):
             if module_name.startswith("cython." + wrong):
                 extra = "Did you mean 'cython.%s' ?" % correct
                 break
+        if not extra:
+            from .. import Shadow
+            if (submodule in Shadow.__dict__
```

Comment:
I don't think we're using the `Shadow` module anywhere else yet, although the namespace doesn't seem wrong here. I'd rather not use something as broad as `Shadow.__dict__` for this comparison, though. I always considered the `Shadow` module a pure user-side thing that we could even pass around as a simple `cython.py` module. Importing it as part of the compiler run feels a bit wrong.

Can't we use information that we already have available in the `ParseTreeTransforms` module (which is where the `cython.*` names are being handled)?

Reply:
There's quite a few sources of information. The fused types like `floating` seem like they're defined in CythonScope.py. I'm not sure off the top of my head where `double`, `int`, etc. come from (but it doesn't look to be `ParseTreeTransforms`.

I also agree that it seems like a bit of an abuse of `Shadow`, but it seemed like the most complete thing we had.

Would `hasattr` be a little less extreme that `__dict__`? Possibly with a bit of filtering for attributes starting with `__`?

Reply:
+1 for `hasattr`